pkg/kubelet/cm/fake_cgroup_manager.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cm
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+type FakeCgroupManager struct {
+	sync.Mutex
+	CalledFunctions []string
+}
+
+func NewFakeCgroupManager() *FakeCgroupManager {
+	return &FakeCgroupManager{}
+}
+
+func (cm *FakeCgroupManager) Create(*CgroupConfig) error {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Create")
+	return nil
+}
+
+func (cm *FakeCgroupManager) Destroy(*CgroupConfig) error {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Destroy")
+	return nil
+}
+
+func (cm *FakeCgroupManager) Update(*CgroupConfig) error {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Update")
+	return nil
+}
+
+func (cm *FakeCgroupManager) Validate(name CgroupName) error {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Validate")
+	return nil
+}
+
+func (cm *FakeCgroupManager) Exists(name CgroupName) bool {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Exists")
+	return false
+}
+
+func (cm *FakeCgroupManager) Name(name CgroupName) string {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Name")
+	return name.ToCgroupfs()
+}
+
+func (cm *FakeCgroupManager) CgroupName(name string) CgroupName {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "CgroupName")
+	return ParseCgroupfsToCgroupName(name)
+}
+
+func (cm *FakeCgroupManager) Pids(name CgroupName) []int {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "Pids")
+	return []int{}
+}
+
+func (cm *FakeCgroupManager) ReduceCPULimits(cgroupName CgroupName) error {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "ReduceCPULimits")
+	return nil
+}
+
+func (cm *FakeCgroupManager) MemoryUsage(name CgroupName) (int64, error) {
+	cm.Lock()
+	defer cm.Unlock()
+	cm.CalledFunctions = append(cm.CalledFunctions, "MemoryUsage")
+	return 0, nil
+}
+
+func (cm *FakeCgroupManager) AssertCalls(calls []string) error {
+	cm.Lock()
+	defer cm.Unlock()
+
+	if !reflect.DeepEqual(calls, cm.CalledFunctions) {
```

Comment:
You're using locks above which implies you have races without the locks, yet it records the exact order and then tests the exact order. This is inconsistent; either the locks aren't needed because there's no race, or the order is indeterminate and it's incorrect to depend on the order like this function does.

Reply:
The idea was to follow the existing pattern of making fake implementations thread safe like FakeContainerManager/FakePodContainerManager. It can be removed, locking is not needed for the current usage.

Reply:
I removed the lock.